Feature

nareshnootoo opened the pull request:
Varify the maximum number of digits that are present in the number(Precision),
    and the maximum number of decimal places(Scale)

    :raises Invalid: If the value does not match the provided Precision and Scale.

    >>> s = Schema(Number(precision=6, scale=2))

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +794,44 @@ def __call__(self, v):
+    resource : http://stackoverflow.com/questions/3018758/determine-precision-and-scale-of-particular-number-in-python
+    """
+
+    def __init__(self, precision=None, scale=None, msg=None, max_digits=14):
```
make `max_digit` constant

tusharmakkar08 commented on voluptuous/tests/tests.py:
```diff
@@ -557,3 +557,31 @@ def fn(arg):
         return "hello"
 
     assert_raises(Invalid, fn, 1)
+
```
Add following test cases: 

- [ ]  scale is `None`
- [ ]  precision is `None`
- [ ]  Both are `None`

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +794,44 @@ def __call__(self, v):
+            frac_digits /= 10
+        scale = int(math.log10(frac_digits))
+
+        if magnitude + scale > self.precision or scale > self.scale :
```
Make 2 separate `Invalid` conditions

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +794,44 @@ def __call__(self, v):
+        scale = int(math.log10(frac_digits))
+
+        if magnitude + scale > self.precision or scale > self.scale :
+            raise Invalid(self.msg or "Precision must be smaller than %s, and Scale must be smaller than %s"%(self.precision, self.scale))
```
This will break when `self.precision is None` or `self.scale is None`

nareshnootoo commented:
@tusharmakkar08 I have made some changes as per your suggestion please review

tusharmakkar08 commented on voluptuous/tests/tests.py:
```diff
@@ -557,3 +557,61 @@ def fn(arg):
+
+def test_number_validation_with_valid_number():
+    """ test with Number with valid precision and scale"""
+    schema = Schema({"number" : Number(precision=6, scale=2)})
```
Shouldn't this yield error?

nareshnootoo replied:
In which scenario?

tusharmakkar08 replied:
I mean this scenario should give error and the correct number should be : 1234.00

nareshnootoo replied:
I think we shouldn't be so hard, as 1234 is a valid number, however we can return it with decimal places, what do you think?

tusharmakkar08 replied:
But the point of having a separate `Number` validator is to tackle these kind of issues only. If we want 1234, directly to be accepted then instead of using `Number` we can use `int` which is already supported by voluptuous.

nareshnootoo replied:
Make sense :)

nareshnootoo replied:
But in python if I declare a variable as `a=1.000` it always give me `1.0`
same in case of PHP it return me as `1`
also in general 1.00 == 1

tusharmakkar08 replied:
You can try some different implementation. Something on the lines of using  [Decimal](https://docs.python.org/2/library/decimal.html) and string conversions.

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +794,60 @@ def __call__(self, v):
+
+class Number(object):
+    """
+    Varify the maximum number of digits that are present in the number(Precision),
```
*Verify

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +794,60 @@ def __call__(self, v):
+    def __repr__(self):
+        return ('Number(precision=%s, scale=%s, msg=%s)' % (self.precision, self.scale, self.msg))
+
+    def get_precision_scale(self, number):
```
private function

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +794,60 @@ def __call__(self, v):
+        :param number:
+        :return: tuple(precision, scale)
+        """
+        max_digits = 14
```
Better make it default argument in the class.

tusharmakkar08 commented:
Hey @nareshnootoo,

Travis build is failing. Please update the PR.

Thanks

nareshnootoo commented:
Hi @tusharmakkar08 I have implemented the Decimal and string conversions. Please review

tusharmakkar08 commented on voluptuous/tests/tests.py:
```diff
@@ -557,3 +557,68 @@ def fn(arg):
+    else:
+        assert False, "Did not raise Invalid for String"
+
+def test_number_validation_with_invalid_precision_scale():
```
Add a test case with only precision and no scale.

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +793,57 @@ def __call__(self, v):
+    :raises Invalid: If the value does not match the provided Precision and Scale.
+
+    >>> schema = Schema({"number" : Number(precision=6, scale=2)})
+    >>> out_ = schema({"number": '1234.01'})
```
Add an assert for doctests here as well.

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +793,57 @@ def __call__(self, v):
+    def __repr__(self):
+        return ('Number(precision=%s, scale=%s, msg=%s)' % (self.precision, self.scale, self.msg))
+
+    def __get_precision_scale(self, number):
```
Let's make `__get_precision_scale` internal instead of `_get_precision_scale`. Refer http://stackoverflow.com/a/6930223/1799475

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +793,57 @@ def __call__(self, v):
+        :param number:
+        :return: tuple(precision, scale, decimal_number)
+        """
+        from decimal import Decimal
```
Move this import up.

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +793,57 @@ def __call__(self, v):
+
+        try:
+            decimal_num = Decimal(number)
+        except:
```
Please avoid having generic Exception catching.

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -806,7 +806,7 @@ class Number(object):
     >>> schema = Schema(Number(precision=6, scale=2))
     >>> schema('1234.01')
-    1234.01
+    Decimal('1234.01')
```
Assert here

nareshnootoo commented:
Hi @tusharmakkar08 I have fixed the following:
1) Moved Decimal import up
2) Added InvalidOperation exception instead of generic
3) Made __get_precision_scale to internal as _get_precision_scale
4) Added assert for doctests
5) Added case with only invalid precision and no scale

tusharmakkar08 commented on voluptuous/validators.py:
```diff
@@ -794,3 +794,57 @@ def __call__(self, v):
+
+    >>> schema = Schema(Number(precision=6, scale=2))
+    >>> schema('1234.01')
+    Decimal('1234.01')
```
Add assert here.

tusharmakkar08 replied:
Can we do something about Decimal output ? (Maybe have string output). @alecthomas What do you think about this?

nareshnootoo replied:
Yes we can have string output, but user can use processed Decimal.

tusharmakkar08 replied:
@nareshnootoo  Let's add a flag which tells what to yield, where default being the string since we are giving input as string only.

tusharmakkar08 commented on voluptuous/tests/tests.py:
```diff
@@ -557,3 +557,78 @@ def fn(arg):
+    assert_equal(float(out_.get("number")), 123456789012.00)
+
+def test_number_when_precision_none_n_invalid_scale():
+    """ test with Number with no precision and invalid scale"""
```
It's difficult to keep track of all the function names in tests since few of them are like `test_number_when_precision_none_n_invalid_scale` while others are like `test_number_when_scale_none_n_valid_precision`. Please follow a consistent convention over here, preferably add precision first and then scale.

tusharmakkar08 commented on voluptuous/tests/tests.py:
```diff
@@ -557,3 +557,101 @@ def fn(arg):
+        assert False, "Did not raise Invalid for String"
+
+
+def test_number_validation_with_valid_precision_scale_yield_decimal_none():
```
yield_decimal_false

nareshnootoo replied:
@tusharmakkar08 there is already a test case named test_number_validation_with_valid_precision_scale_yield_decimal_false

tusharmakkar08 replied:
I mean add for yield_decimal_true since this test is same as test at Line 652.

tusharmakkar08 commented:
Hey @nareshnootoo 

The commits haven't been squashed (Also fix the comment in my last review). After doing rebase, you might need to do `git push -f` since if you will pull, it will pull all the old commits as well. 

Thanks

nareshnootoo commented:
@tusharmakkar08 I have squashed the commits.

tusharmakkar08 commented:
@nareshnootoo : Please check the earlier comment.

tusharmakkar08 commented:
Hey @nareshnootoo 

Please squash the changes and get this merged into master 🍰 

Thanks

nareshnootoo commented:
Hi @tusharmakkar08 I have squashed the commits.

tusharmakkar08 commented:
Merged 🍰

nareshnootoo commented:
Thanks :)
